pkg/storage/stores/tsdb/multitenant.go

```diff
@@ -66,7 +67,18 @@ func (m *MultiTenantIndex) Series(ctx context.Context, userID string, from, thro
 }
 
 func (m *MultiTenantIndex) LabelNames(ctx context.Context, userID string, from, through model.Time, matchers ...*labels.Matcher) ([]string, error) {
-	return m.idx.LabelNames(ctx, userID, from, through, withTenantLabel(userID, matchers)...)
+	res, err := m.idx.LabelNames(ctx, userID, from, through, withTenantLabel(userID, matchers)...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Strip out the tenant label in response.
+	i := sort.SearchStrings(res, TenantLabel)
+	if i == len(res) || res[i] != TenantLabel {
+		return res, nil
+	}
+
+	return append(res[:i], res[i+1:]...), nil
 }
 
 func (m *MultiTenantIndex) LabelValues(ctx context.Context, userID string, from, through model.Time, name string, matchers ...*labels.Matcher) ([]string, error) {
```

Comment:
double checking: shouldn't we have a similar check for `labelValues`?

Reply:
Good question!
I left it out for speed benefits and because it's unlikely to be queried as it's never returned by the `series` or `labelnames` calls. It should be possible to get a response for this value by querying `labelvalues(LokiTenantLabel)`, but it's unusable outside of that since it's a synthetic label that's only used for packing multiple tenants into a single TSDB pre-compaction.

It _is_ technically correct to remove it though :)

Reply:
followup: it's probably negligible in latency benefits, so I should remove it. Will do!